### PR TITLE
NetGen CSG Support

### PR DESCRIPTION
**Pull request by UZerbinati**
I've added a small function that allows using NetGen mesh created in Python using the NetGen CSG tools. I've also added a very tiny demo.

### CODE REVIEW

**Review by dham — CHANGES_REQUESTED** [review]
Thanks for your first contribution to Firedrake! It would be really nice to have netgen integration as you propose. There is a bit of work to do to get this into mergable  state, but it's a good start. I have left off commenting on the demo until we get the actual code in shape, so start by going down those options.

We'll also need to put in some simple tests of the functionality so that we'll know if something breaks this code in the future. The best thing to do is to add in a new test file to `firedrake/tests/regression` called something like `test_netgen.py`. What you'll want to do is make a simple mesh (maybe just a square) and then conduct convergence tests to a known analytic solution using the method of manufactured solutions. `firedrake/tests/regression/test_helmholtz.py` is a useful model, however since boundary conditions are an important part of netgen functionality, I suggest you use the poisson equation with a Dirichlet BC instead of Helmholtz with natural conditions. For more on MMS see: https://finite-element.github.io/6_finite_element_problems.html#the-method-of-manufactured-solutions

You can run your test locally by running:
```
pytest tests/regression/test_netgen.py
```
We'll have to make sure the build testers have netgen installed, but we can deal with that later.

As a final note, please `make lint` and fix all the style errors it finds.

Thanks again!

File: firedrake/mesh.py [hunk 1]
```diff
@@ -224,6 +224,36 @@ def facet_cell_map(self):
...
+    Create a DMPlex from an Netgen mesh
+
+    :arg ngmesh: NetGen Mesh
+    """
+    if len(ngmesh.Elements3D()) != 0:
+        raise RuntimeError("Only 2D meshes are implemented at the moment.")
+    else:
+        V = [];
+        T = [];
+        for v in list(ngmesh.Points()):
```
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py:239]: I was using an old version of NetGen and it didn't have the Coordinates function implemented yet, I'm gonna use this function.

File: firedrake/mesh.py [hunk 2]
```diff
@@ -224,6 +224,36 @@ def facet_cell_map(self):
...
+            V = V + [[float(v[0].__repr__()),float(v[1].__repr__())]]
+        for t in ngmesh.Elements2D():
+            T = T + [[int(v.__repr__())-1 for v in t.vertices]]
```
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py:242]: I didn't know I could have used `v.nr`. I've implemented this suggestion. A possible way of avoiding the loop over all `Elements2D` could be to expose using pybind11 `Mesh::SurfaceElements`, I'll fork the Netgen repo and see what I can do :)

File: firedrake/mesh.py [hunk 3]
```diff
@@ -224,6 +224,36 @@ def facet_cell_map(self):
...
+    plex = _from_cell_list(2, T, V, comm, name=None)
+    for e in ngmesh.Elements1D():
+        vStart, vEnd = plex.getDepthStratum(0)   # vertices
+        join = plex.getJoin([vStart+int(v.__repr__())-1 for v in e.vertices])
+        plex.setLabelValue(dmcommon.FACE_SETS_LABEL, join[0], int(e.index))
+    return plex
+@PETSc.Log.EventDecorator()
+def ngLabels(ngmesh,labels):
```
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py:250]: So sorry, but I don't understand what you mean by a Netgen functionality.
Netgen allows you to use string labels at the boundary and this little function allows you to specify the same labels used when defining the geometry via Netgen to impose boundary conditions.

File: firedrake/mesh.py [hunk 4]
```diff
@@ -224,6 +224,36 @@ def facet_cell_map(self):
...
 
 
+@PETSc.Log.EventDecorator()
+def _from_netgen(ngmesh,comm=None):
+    """
+    Create a DMPlex from an Netgen mesh
+
+    :arg ngmesh: NetGen Mesh
+    """
+    if len(ngmesh.Elements3D()) != 0:
```
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py:234]: I only thought about the 2D implementation, but I'll have a look also to the 3D case. There should be no major difference as you were pointing out.
I've never tried to do a 1D mesh in Netgen, but I think they are supported. Do you think it would be useful to have 1D mesh implemented?

File: firedrake/mesh.py [hunk 5]
```diff
@@ -1906,7 +1936,7 @@ def Mesh(meshfile, **kwargs):
 
     """
     import firedrake.function as function
-
+
```
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py:1939]: I added minimal comments in the docstring, but I don't know if they are sufficient.

**Review by ksagiyam — CHANGES_REQUESTED** [review]

File: firedrake/mesh.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ksagiyam) [reply firedrake/mesh.py]: So, as far as I understand, `ngmesh.GetBCName(ID)` maps the boundary ID (`ID`) to the boundary name; it would be a lot nicer if we had the inverse of this map (say `ngmesh.GetBCIDs("boundary_name")`-> list of boundary IDs) available in Netgen. (As for current code, we should `unique(BID)`.

File: firedrake/mesh.py [hunk 6]
```diff
@@ -1936,6 +1971,8 @@ def Mesh(meshfile, **kwargs):
     geometric_dim = kwargs.get("dim", None)
     if isinstance(meshfile, PETSc.DMPlex):
         plex = meshfile
+    elif netgen and isinstance(meshfile, netgen.libngpy._meshing.Mesh):
+        plex = _from_netgen(meshfile, comm)
```
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py:1987]: What do you mean for name ?

File: firedrake/mesh.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py]: Ah, now I get what you mean. I'm sorry I haven't yet had time to have a look at the Netgen source code. I'll try to add this feature next week.

File: firedrake/mesh.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ksagiyam) [reply firedrake/mesh.py:1987]: Name of the mesh (such as `my_mesh_A` or `cylinder_mesh_B`), which we need in order to save the mesh in file at later stage if needed. When we call this `Mesh` function, `name` is passed as an argument (otherwise the default value is assigned to `name`); I thought the Netgen mesh object would carry some name with it, but it seemed not.

[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py:1987]: Ah, I'll have a look into it. Maybe it is possible to add this label on the Netgen side.

File: firedrake/mesh.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py]: Ok, now it should be a bit better, but in all fairness, I just moved the loops on the C++ side of the Netgen wrapper. This seems to be speeding up things significantly, but probably the best way to do it would be to use the memoryview feature from Pybind11.

File: firedrake/mesh.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py]: Added to netgen GetBCIDs, on my [fork](https://github.com/UZerbinati/netgen). I'll wait to add all the features needed for the firedrake interface and then open a pull request.

File: firedrake/mesh.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py:1976]: I've added this feature in my Netgen fork, now there are a `SetMeshName(name)` and `GetMeshName(name)` functions.

File: firedrake/mesh.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py]: @dham I think I've implemented all yours and @ksagiyam suggestions for the 2D case, I'll try to have a look at the 3D case soon.

[the remark this answers is not in the document]
Comment (UZerbinati) [reply firedrake/mesh.py]: Added 3D support but it requires a function from PETSc to be wrapped in petsc4py, I've done it in [this fork](https://gitlab.com/UZerbinati1/petsc/-/tree/uzerbianti/petsc4py).

**Review by UZerbinati — COMMENTED** [review]
I've added a couple of functions in Netgen in order to make the creation of mesh using PETSc DMPlex easier. All modifications can be found in [this fork](https://github.com/UZerbinati/netgen) and I created a corresponding [PR](https://github.com/NGSolve/netgen/pull/134).
I've also exposed in petsc4py the getFulJoin method. This can be found in [this fork](https://gitlab.com/UZerbinati1/petsc/-/tree/uzerbianti/petsc4py) and I created a corresponding [PR](https://gitlab.com/petsc/petsc/-/merge_requests/5558).

File: firedrake/mesh.py [hunk 7]
```diff
@@ -224,6 +229,41 @@ def facet_cell_map(self):
...
+
+    :arg ngmesh: NetGen Mesh
+    """
+    if ngmesh.dim == 3:
+        V = ngmesh.Coordinates()
+        T = ngmesh.Indices3D
+        plex = _from_cell_list(3, T, V, comm, name=None)
+        plex.setName(_generate_default_mesh_topology_name(ngmesh.GetMeshName()))
+        for e in ngmesh.Elements2D():
+            vStart, vEnd = plex.getDepthStratum(0)   # vertices
```
Comment (ksagiyam) [inline firedrake/mesh.py:245]: Can move this line outside the loop.

File: firedrake/mesh.py [hunk 8]
```diff
@@ -224,6 +229,41 @@ def facet_cell_map(self):
...
+            join = plex.getFullJoin([vStart+v.nr-1 for v in e.vertices])
+            plex.setLabelValue(dmcommon.FACE_SETS_LABEL, join[0], int(e.index))
+        for e in ngmesh.Elements1D():
+            vStart, vEnd = plex.getDepthStratum(0)   # vertices
```
Comment (ksagiyam) [inline firedrake/mesh.py:249]: Then can remove this line.

File: firedrake/mesh.py [hunk 9]
```diff
@@ -224,6 +229,41 @@ def facet_cell_map(self):
...
+            join = plex.getJoin([vStart+v.nr-1 for v in e.vertices])
+            plex.setLabelValue(dmcommon.FACE_SETS_LABEL, join[0], int(e.index))
```
Comment (ksagiyam) [inline firedrake/mesh.py:251]: I think this can be `dmcommon.EDGE_SETS_LABEL` after adding the following line at the beginning of `dmcommon.pyx`
```
EDGE_SETS_LABEL = "Edge Sets"
```
"Edge Sets" label does not exist in PETSc, but I don't think it does harm adding it in Firedrake.

**Review by ksagiyam — CHANGES_REQUESTED** [review]
Thanks. I hope your Netgen PR will be merged soon.

File: firedrake/mesh.py (hunk 9 above)
Comment (UZerbinati) [reply firedrake/mesh.py]: Done :) Btw If I'd like also to assign a material ID to simplices should I use the CELL_SETS_LABEL?

File: firedrake/mesh.py [hunk 10]
```diff
@@ -1868,7 +1907,7 @@ def Mesh(meshfile, **kwargs):
     Meshes may either be created by reading from a mesh file, or by
     providing a PETSc DMPlex object defining the mesh topology.
 
-    :param meshfile: Mesh file name (or DMPlex object) defining
+    :param meshfile: Colud be the mesh file name, a DMPlex object or a Netgen mesh object defining
```
Comment (dham) [inline firedrake/mesh.py:1910]: Just delete "Could be", it adds no information.

File: firedrake/mesh.py (hunk 9 above)
Comment (dham) [reply firedrake/mesh.py]: Yes, that's right.

File: demos/netgen/NetGenPoisson.rst [hunk 11]
```diff
@@ -0,0 +1,89 @@
+We import the needed library from FireDrake and NetGen.
```
[the remark this answers is not in the document]
Comment (dham) [reply demos/netgen/NetGenPoisson.rst:1]: OK now that we're just about there with the functionality, let's talk about the demo. 

The demo file needs to have the extension `.py.rst`, and the convention for the base name is that it should be lower case with words separated by underscores. So `netgen_poisson.py.rst` would be a logical name in this case.

Firedrake is one word and doesn't have a capital D.

For the demo to be useful for users, it needs to have quite a bit more text. Take a look at `demos/immersed_fem/immersed_fem.py.rst` for an example of the level of detail that is good. In particular, there should be a title and a couple of sentences of introduction, presumably linking to the Netgen Python interface docs.

You should test that the demo works by running `make demos` in the Firedrake source directory to generate `netgen_poisson.py` and then running the resulting file.

Some more detailed comments are provided below.

File: demos/netgen/NetGenPoisson.rst [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+.. code:: ipython3
+
+    from firedrake import *
+    import firedrake.cython.dmcommon as dmcommon
+    from firedrake import mesh
+    from ngsolve import Mesh,Draw
+    from netgen.geom2d import CSG2d, Circle, Rectangle
```
Comment (dham) [inline demos/netgen/NetGenPoisson.rst:9]: make sure you don't have redundant imports. For example likes 6 and 7 are definitely not needed.

File: demos/netgen/NetGenPoisson.rst [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+we define a geometry using NetGen constructive geometry tools.
+
+.. code:: ipython3
+
+    #EXAMPLE OF Constructive Solid Geometry
```
Comment (dham) [inline demos/netgen/NetGenPoisson.rst:15]: Prefer rst text outside code blocks to comments. You can have as many code blocks as are needed to explain what is going on.

File: demos/netgen/NetGenPoisson.rst [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    u = TrialFunction(V)
+    v = TestFunction(V)
+    
+    f = Function(V)
+    x, y = SpatialCoordinate(msh)
+    f.interpolate(1+0*x)
+    a = inner(grad(u), grad(v)) * dx
+    l = inner(f, v) * dx
+    
+    print("Problem Seted Up !")
```
Comment (dham) [inline demos/netgen/NetGenPoisson.rst:65]: Don't put progress print statements through the demo. They don't really add any information.

File: demos/netgen/NetGenPoisson.rst [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    
+    #For CSG Example use [1, 2, 3, 10, 11, 12] as BID
+    bc = DirichletBC (V , 0.0 , mesh.ngLabels(ngmesh,["inlet"])) # Boundary condition
+    print("Assembling ...")
+    A = assemble (a , bcs = bc )
+    b = assemble (l)
+    bc.apply(b)
+    print("Assembled !")
+    print("Solving ...");
+    solve (A, u, b, solver_parameters ={"ksp_type": "preonly", "pc_type": "lu"})
```
Comment (dham) [inline demos/netgen/NetGenPoisson.rst:77]: These are the default solver parameters so you don't need to provide them.

File: demos/netgen/NetGenPoisson.rst [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    File("Poisson.pvd").write(u)
+
+
+.. parsed-literal::
+
+    Problem Seted Up !
+    Assembling ...
+    Assembled !
+    Solving ...
+    Solved !
```
Comment (dham) [inline demos/netgen/NetGenPoisson.rst:88]: As above, plotting the solution would give users much more information than showing the progress prints.

**Review by dham — CHANGES_REQUESTED** [review]
Thanks again for this, we're nearly there now. I've gone through and given some pointers on how to finish the demo.

**Comment by UZerbinati** [comment]
Dear @dham,
I'm also trying to use the mesh refinement function included in NetGen  (I've exposed the function needed in my  NetGen fork) and include this in the documentation. Is there a way given a function to find out to what mesh cell the degree of freedom of the dat.data vector corsipond to ?

**Comment by ksagiyam** [comment]
`f.function_space().cell_node_list` maps Firedrake cell numbers to function space nodes:
https://github.com/firedrakeproject/firedrake/blob/ee720d61991e7f04004847dce79844771fad6e2b/firedrake/functionspaceimpl.py#L450
So what you want is basically the inverse of this map. What do you need this for?

**Comment by UZerbinati** [comment]
> 
Dear @ksagiyam,
I need it in the refinement loop. I compute the error indicator using the DG space of order 0 and then refine only the cell where the corresponding DOF is greater than a certain tolerance. I attach an example script.
[ell-doug-eig.txt](https://github.com/firedrakeproject/firedrake/files/9636278/ell-doug-eig.txt)

**Comment by UZerbinati** [comment]
> > 
> 
> Dear @ksagiyam, I need it in the refinement loop. I compute the error indicator using the DG space of order 0 and then refine only the cell where the corresponding DOF is greater than a certain tolerance. I attach an example script. [ell-doug-eig.txt](https://github.com/firedrakeproject/firedrake/files/9636278/ell-doug-eig.txt)

Ok I found out the problem is the difference between the cell numbering between NetGen and the PETSc DMPLEX, I'll think about how to create a map between the two. Right now I have a very naive implementation that seems to do the trick.

**Comment by ksagiyam** [comment]
Please note that Firedrake cell numbers are different from DMPlex cell numbers; those are related by `mesh._cell_numbering`. If you are only interested in mapping Firedrake cell numbers to DMPlex cell number, you can use `mesh.cell_closure[:,-1]`.

**Comment by rckirby** [comment]
A few questions/comments:
- Demo on how to tag different parts of the boundary with different boundary labels would be helpful (e.g. Dirichlet BC on left and right ends, Neuman on top and bottom of a box)
- NGSolve does multigrid, IIUC, so is it possible for us to define a netgen CSG geometry and get a (possibly non-nested) multigrid hierarchy?
- Have you checked that this works in complex arithmetic (mesh coordinates are stored as complex numbers in that case...)
- Can we get higher-order/curvilinear geometry out of this?

**Comment by UZerbinati** [comment]
> A few questions/comments:
> 
> * Demo on how to tag different parts of the boundary with different boundary labels would be helpful (e.g. Dirichlet BC on left and right ends, Neuman on top and bottom of a box)
I'm writing more consistent documentation at the moment. I'll definitely include that!
> * NGSolve does multigrid, IIUC, so is it possible for us to define a netgen CSG geometry and get a (possibly non-nested) multigrid hierarchy?
I've not yet looked into it, but I put this on my to-do list. I've been experimenting with mesh refinement using NetGen and Firedrake for domains with corner singularity (I'm also including this in the new demo). There you can get child and parent refinement out of the NetGen mesh quite easily.
> * Have you checked that this works in complex arithmetic (mesh coordinates are stored as complex numbers in that case...)
No, I never tested that. I'll put it in my todo list :)
> * Can we get higher-order/curvilinear geometry out of this?
Yes, you can in particular if you are using an OCC geometry `MakeFillet` method. Most of the documentation on this kind of thing is contained [here](https://docu.ngsolve.org/latest/i-tutorials/index.html#geometric-modeling-and-mesh-generation). But I must say I don't know how Firedrake would treat high-order/curvilinear elements, this is something I was hoping to have a look into next.

**Comment by rckirby** [comment]
> Dear @rckirby,
> 
> > A few questions/comments:
> > 
> > * Demo on how to tag different parts of the boundary with different boundary labels would be helpful (e.g. Dirichlet BC on left and right ends, Neuman on top and bottom of a box)
> 
> I'm writing more consistent documentation at the moment. I'll definitely include that!

Thanks, very good.

> 
> > * NGSolve does multigrid, IIUC, so is it possible for us to define a netgen CSG geometry and get a (possibly non-nested) multigrid hierarchy?
> 
> I've not yet looked into it, but I put this on my to-do list. I've been experimenting with mesh refinement using NetGen and Firedrake for domains with corner singularity (I'm also including this in the new demo). There you can get child and parent refinement out of the NetGen mesh quite easily.
> 
> > * Have you checked that this works in complex arithmetic (mesh coordinates are stored as complex numbers in that case...)
> 
> No, I never tested that. I'll put it in my todo list :)
> 
It's probably an easy fix if it doesn't (grab `.real` of the coordinates).
> > * Can we get higher-order/curvilinear geometry out of this?
> 
> Yes, you can in particular if you are using an OCC geometry `MakeFillet` method. Most of the documentation on this kind of thing is contained [here](https://docu.ngsolve.org/latest/i-tutorials/index.html#geometric-modeling-and-mesh-generation). But I must say I don't know how Firedrake would treat high-order/curvilinear elements, this is something I was hoping to have a look into next.

You might look at the current `OpenCascadeMeshHierarchy` for curvilinear and mesh hieararchies.  I find it very fussy to use (often breaks/gives bad meshes for a problem I'm working on/drops BC labels on the floor), but it aims to do about what I want.  I'd be very interested if somehow your hookup were more robust with BC labeling and some weird geometry but could still give me hieararchies to do MG with.
